packages/nodejs/src/__tests__/transmitter.test.ts

```diff
@@ -114,6 +105,165 @@ describe("Transmitter", () => {
+
+    it("resolves to a response stream on success", async () => {
+      nock("http://example.com").get("/foo").reply(200, "response body")
+
+      const stream = await transmitter.download()
+
+      await expect(
+        new Promise(resolve => {
+          stream.on("data", resolve)
+        })
+      ).resolves.toEqual(Buffer.from("response body"))
+    })
+
+    it("rejects if the status code is not successful", async () => {
+      nock("http://example.com").get("/foo").reply(404, "not found")
+
+      await expect(transmitter.download()).rejects.toMatchObject({
+        kind: "statusCode",
+        statusCode: 404
+      })
+    })
+
+    it("rejects if there's a request error", async () => {
+      // do not mock the request -- beforeEach uses nock to disable network
+      // requests, so any request to a non-mocked host will throw an error
+
+      await expect(transmitter.download()).rejects.toMatchObject({
+        kind: "requestError",
+        error: expect.any(Error)
+      })
+    })
+  })
+
+  describe(".request", () => {
+    function mockRequest(module: typeof http | typeof https) {
+      const requestCallbacks = {
+        on: jest.fn(),
+        write: jest.fn(),
+        end: jest.fn()
+      }
+
+      // casting the spy to `any` because the type system gets
+      // confused about the expected `mockImplementation` signature
+      const spy: any = jest.spyOn(module, "request")
+
+      const mock = spy.mockImplementation(
+        (_options: any, callback: (stream: any) => void) => {
+          const stream = new EventEmitter()
+          callback(stream)
+          stream.emit("end")
+
+          return (requestCallbacks as unknown) as ClientRequest
+        }
+      )
+
+      return mock
+    }
+
+    function transmitterRequest(
+      method: string,
+      url: string,
+      body?: string,
+      params = new URLSearchParams()
+    ) {
+      return new Promise<HashMap<jest.Mock>>(resolve => {
+        const callbacks: HashMap<jest.Mock> = {
+          callback: jest.fn(stream => {
+            stream.on("data", callbacks.onData)
+            stream.on("end", () => resolve(callbacks))
+          }),
+
+          onData: jest.fn(),
+          onError: jest.fn(() => resolve(callbacks))
+        }
+
+        const transmitter = new Transmitter(url, body)
+
+        transmitter.request({
+          method,
+          params,
+          callback: callbacks.callback,
+          onError: callbacks.onError
+        })
+      })
+    }
+
+    it("performs an HTTP GET request", async () => {
+      nock("http://example.invalid").get("/foo").reply(200, "response body")
+
+      const { callback, onData, onError } = await transmitterRequest(
+        "GET",
+        "http://example.invalid/foo"
+      )
+
+      expect(callback).toHaveBeenCalledWith(
+        expect.objectContaining({ statusCode: 200 })
+      )
+
+      expect(onData).toHaveBeenCalledWith(Buffer.from("response body"))
```

Comment:
Do we need to test this (the response body and status code), on what is basically a private function, if the transmit and download function tests also already tests this?

Reply:
On this comment, and the one below about making it private: I've been thinking of this refactor as going in the opposite direction. I think that, if we implemented a good enough `.request` method, we could get rid of the `.transmit` and `.download` helper methods entirely, and just call `.request` with the needed parameters whenever we need to perform these requests.

I think it's getting there, but for now the caller still needs to provide a callback that figures out what to do with the response object, and it still needs to wrap the whole thing in a promise itself. If the `.request` method took care of that, too, then using it would be fairly straightforward, and the helpers would be unnecessary.

Reply:
That sounds good. I'm not sure if we need to update that right now. Let's focus on some other things first. Can you create an issue with your plans for a refactor?